ssl/ssl_lib.c

```diff
@@ -6901,3 +6909,173 @@ int SSL_CTX_set0_tmp_dh_pkey(SSL_CTX *ctx, EVP_PKEY *dhpkey)
     ctx->cert->dh_tmp = dhpkey;
     return 1;
 }
+
+#ifndef OPENSSL_NO_COMP
+
+static int ssl_set_cert_comp_pref(uint16_t *array, size_t array_size, uint16_t *algs, int num_algs)
+{
+    int i, j = 0;
+
+    /* This will 0-terminate the array */
+    memset(array, 0, array_size);
+    /* include only those algorithms we support, ignore the rest if the list fills */
+    for (i = 0; i < num_algs && j < TLSEXT_comp_cert_limit; i++) {
+#if defined(BROTLI)
+        if (algs[i] == TLSEXT_comp_cert_brotli)
+            array[j++] = algs[i];
+#endif
+#if defined(ZLIB)
+        if (algs[i] == TLSEXT_comp_cert_zlib)
+            array[j++] = algs[i];
+#endif
+#if defined(ZSTD)
+        if (algs[i] == TLSEXT_comp_cert_zstd)
+            array[j++] = algs[i];
+#endif
+    }
+    return 1;
+}
+#endif
+
+int SSL_CTX_set_cert_comp_preference(SSL_CTX *ctx, uint16_t *algs, int num_algs)
+{
+#ifndef OPENSSL_NO_COMP
+    return ssl_set_cert_comp_pref(ctx->cert_comp_prefs, sizeof(ctx->cert_comp_prefs),
+                                  algs, num_algs);
+#else
+    return 1;
+#endif
+}
+
+int SSL_set_cert_comp_preference(SSL *ssl, uint16_t *algs, int num_algs)
+{
+#ifndef OPENSSL_NO_COMP
+    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_SSL(ssl);
+
+    if (sc == NULL)
+        return 0;
+    return ssl_set_cert_comp_pref(sc->cert_comp_prefs, sizeof(sc->cert_comp_prefs),
+                                  algs, num_algs);
+#else
+    return 1;
+#endif
+}
+
+int SSL_CTX_set_cert_comp_cb(SSL_CTX *ctx, OSSL_cert_comp_fn cert_comp_cb, void *arg)
+{
+#ifndef OPENSSL_NO_COMP
+    ctx->cert_comp_cb = cert_comp_cb;
+    ctx->cert_comp_arg = arg;
+#endif
+    return 1;
+}
+
+int SSL_set_cert_comp_cb(SSL *ssl, OSSL_cert_comp_fn cert_comp_cb, void *arg)
+{
+#ifndef OPENSSL_NO_COMP
+    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_SSL(ssl);
+
+    if (sc == NULL)
+        return 0;
+    sc->cert_comp_cb = cert_comp_cb;
+    sc->cert_comp_arg = arg;
+#endif
+    return 1;
+}
+
+unsigned char *SSL_get_cert_to_compress(SSL *ssl, size_t *length)
+{
+#ifndef OPENSSL_NO_COMP
+    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_SSL(ssl);
+    WPACKET tmppkt;
+    BUF_MEM buf = { 0 };
+    unsigned char *ret = NULL;
+
+    if (sc == NULL || length == NULL)
+        return NULL;
+
+    if (!WPACKET_init(&tmppkt, &buf)
+            || sc->ext.compress_certificate_from_peer == TLSEXT_comp_cert_none)
+        goto err;
+
+    /* Use the |tmppkt| for the to-be-compressed data */
+    if (sc->server) {
+        /* no context present, add 0-length context */
+        if (!WPACKET_put_bytes_u8(&tmppkt, 0))
+            goto err;
+        if (!ssl3_output_cert_chain(sc, &tmppkt, sc->s3.tmp.cert)) {
+            /* SSLfatal() already called */
```

Comment:
Yup,  It shouldn't sent an alert/fatal error; this is all here because `ssl3_output_cert_chain()` calls `SSLfatal()`. Which I probably need to tweak to not do in this case.

Reply:
However, if it's erroring out, it will error out without compression, as the same sequence of APIs (`ssl3_output_cert_chain()`) will still end up being called. So, this one's a bit tricky, and behavior may be different if called outside the callback.

Reply:
This was fixed using an `SSL_CONNECTION` flag to avoid sending an alert, and clearing out the errors before return. Not sure there's an easier/cleaner way to do this without copying a ton of code (and subsequently ossifying the copy) or adding a ton of conditionals.